### PR TITLE
Do not start the runner script if it's already running

### PR DESCRIPTION
**Pull request by enescakir**
In the "register_runner" label, if the respirate crashes after the "./actions-runner/run.sh" script but before it hops, it fails in the subsequent run. If it has already started, we can hop to the "wait" label. The "wait" label determines the next course of action if the service fails to start.

If the runner-script service fails, we register the runner again. However, to do this, we must clean up the runner configurations and the systemd service.

### CODE REVIEW

File: prog/vm/github_runner.rb [hunk 1]
```diff
@@ -189,14 +190,24 @@ def before_run
...
       github_runner.incr_destroy
       nap 0
     when "failed"
+      Clog.emit("#{SERVICE_NAME} is failed, registering the runner again") { {github_runner: github_runner.values} }
+      vm.sshable.cmd("sudo systemctl stop runner-script")
+      vm.sshable.cmd("./actions-runner/config.sh remove --local")
       github_client.delete("/repos/#{github_runner.repository_name}/actions/runners/#{github_runner.runner_id}")
       github_runner.update(runner_id: nil, ready_at: nil)
       hop_register_runner
+    when "dead"
```
Comment (furkansahin) [inline prog/vm/github_runner.rb:209]: is it possible for this to exhaust all of the VMs in the pool? Let's assume there is a bug and the script is constantly failing, would that cause a continuous VM creation/deletion?

File: prog/vm/github_runner.rb [hunk 2]
```diff
@@ -188,14 +189,24 @@ def before_run
...
+    #   running => InProgress
+    #   exited  => Succeeded
+    #   failed  => Failed
     case vm.sshable.cmd("systemctl show -p SubState --value #{SERVICE_NAME}").chomp
     when "exited"
       github_runner.incr_destroy
       nap 0
     when "failed"
+      Clog.emit("#{SERVICE_NAME} is failed, registering the runner again") { {github_runner: github_runner.values} }
+      vm.sshable.cmd("sudo systemctl stop runner-script")
```
Comment (byucesoy) [inline prog/vm/github_runner.rb:203]: For failures, you need to use `sudo systemctl reset-failed runner-script.service`. (see the daemonizer)

Also we use SERVICE_NAME in some places and runner-script in some others. I think you don't need SERVICE_NAME constant. You can use runner-script directly.

File: prog/vm/github_runner.rb [hunk 3]
```diff
@@ -155,6 +155,9 @@ def before_run
   end
 
   label def register_runner
+    # Do not register runner if the runner-script service is already running.
```
Comment (byucesoy) [inline prog/vm/github_runner.rb:158]: Why would that happen? wait ensures that the service is not running before redirecting to here.

File: prog/vm/github_runner.rb [hunk 4]
```diff
@@ -188,14 +189,24 @@ def before_run
...
       github_runner.incr_destroy
       nap 0
     when "failed"
+      Clog.emit("#{SERVICE_NAME} is failed, registering the runner again") { {github_runner: github_runner.values} }
+      vm.sshable.cmd("sudo systemctl stop runner-script")
+      vm.sshable.cmd("./actions-runner/config.sh remove --local")
       github_client.delete("/repos/#{github_runner.repository_name}/actions/runners/#{github_runner.runner_id}")
       github_runner.update(runner_id: nil, ready_at: nil)
       hop_register_runner
+    when "dead"
```
Comment (byucesoy) [inline prog/vm/github_runner.rb:208]: Service cannot be not started yet at this point, no?

**Comment by fdr** [comment]
Tagging out since everyone else seems to be deep-reading this one. I wish I could say I have all the symbols returned from systemd completely memorized. Or...do I wish that?

File: prog/vm/github_runner.rb (hunk 1 above)
Comment (enescakir) [reply prog/vm/github_runner.rb]: I removed this case

File: prog/vm/github_runner.rb (hunk 2 above)
Comment (enescakir) [reply prog/vm/github_runner.rb]: Changed it

File: prog/vm/github_runner.rb (hunk 4 above)
Comment (enescakir) [reply prog/vm/github_runner.rb]: Removed it

File: prog/vm/github_runner.rb (hunk 3 above)
Comment (enescakir) [reply prog/vm/github_runner.rb:191]: If the respirate crashes after the "./actions-runner/run.sh" script but before it hops, it fails in the subsequent run. We investigate a page like that with you.

**Comment by enescakir** [comment]
> Tapping out since everyone else seems to be deep-reading this one. I wish I could say I have all the symbols returned from systemd completely memorized. Or...[do I wish that? ](https://en.wikipedia.org/wiki/The_Monkey%27s_Paw)

Thanks to @byucesoy's daemonizer 😄 I just copied states from there.

https://github.com/ubicloud/ubicloud/blob/7427e66dc8cf092a486f46a5ef3be351e6136c7c/rhizome/common/bin/daemonizer#L12

File: prog/vm/github_runner.rb (hunk 3 above)
Comment (enescakir) [reply prog/vm/github_runner.rb:191]: However, it's extremely rare. I'm uncertain if it's worthwhile to incorporate an additional SSH call for all runs. In 99.99% of cases, it returns "dead".

Comment (byucesoy) [reply prog/vm/github_runner.rb:191]: I think it is OK to have ssh call in this case. We don't do it frequently.

File: prog/vm/github_runner.rb [hunk 5]
```diff
@@ -188,6 +188,9 @@ def before_run
   end
 
   label def register_runner
+    # Do not register runner if the runner-script service is already running.
+    hop_wait unless vm.sshable.cmd("systemctl show -p SubState --value #{SERVICE_NAME}").chomp == "dead"
```
Comment (byucesoy) [inline prog/vm/github_runner.rb:192]: This handles the case where we execute the vm.sshable.cmd but not hop_wait.

How about the case where we execute github_client.post but not the rest. Would we get "Already exist" error and then delete the runner? We shouldn't delete the runner in that case, no?

**Review by byucesoy — APPROVED** [review]

**Pull request closed without merging**